### ISSUE TITLE
Validation of protocol message signature failed (simpleSAMLphp IdP)

### ISSUE THREAD

**Issue by marconett** [bitium/confluence-saml-plugin#29]
Hi,

I'm running into problems setting up the Confluence plugin with simpleSAMLphp as IdP and I'm not sure what's wrong. My setup is very basic.

### Versions
Confluence 5.8.14
simpleSAMLphp 1.13.2
Confluence SAML 2.0 Plugin 1.0.3 (the one from the marketplace)

### My confluence setup

My IdPs Login- and Logout-URL both woth with other service providers.

The plugins docs state, that users must exist prior to logging in, so I created a user named "thomas" in Confluence and configured the plugin like this:

![Plugin Settings](http://i.imgur.com/tCQmMSf.png)

### My simpleSAMLphp setup
I have a user named "thomas" on the IdPs LDAP-backend. Signing thomas into other services providers works fine.

The following is from `metadata/saml20-sp-remote.php` and defines the Confluence SP for the my simpleSAMLphp IdP:

```php
$metadata['http://192.168.100.117:8090/confluenceSAML'] = array(
        'AssertionConsumerService' => 'http://192.168.100.117:8090/plugins/servlet/saml/auth',
        'NameIDFormat' => 'urn:oasis:names:tc:SAML:2.0:attrname-format:basic',
        'authproc' => array(
            1 => array(
                // 'class' => 'saml:PersistentNameID',
                // 'class' => 'saml:AttributeNameID',
                'class' => 'saml:AttributeNameID',
                'attribute' => 'uid',
                'Format' => 'urn:oasis:names:tc:SAML:2.0:attrname-format:basic',
            ),
        ),
```

I read somwhere that the plugins needs the username ("thomas") inside the `NameID` field. With this config, the username is present in plain text inside the IdPs SAML response.

### The SAML-Response from my IdP
http://pastebin.com/iK2jAQ7A (note line 61-63)

### The problem

This is the error message the `atlassian-confluence.log` gives me.

```
2015-10-21 15:20:51,367 ERROR [http-nio-8090-exec-4] [bitium.confluence.servlet.SsoLoginServlet] doPost saml plugin error + Validation of protocol message signature failed
 -- referer: http://192.168.100.117/simplesaml/saml2/idp/SSOService.php?SAMLRequest=fVJdT8JAEPwrl32n9yEgXCgGNUYSDMRWH3w7y2KPtHe1eyX%2BfM%2BCURPD4%2B3N7Mzu7Ozqo67YAVuy3qUgEwEMXeG31r2l8JTfDSZwNZ%2BRqSvV6EUXSveI7x1SYJHoSB9%2FUuhap70hS9qZGkmHQmeLh5VWidBN64MvfAVsQYRtiFI33lFXY5the7AFPj2uUihDaDTncqoSOZ4kUohEyks9EVPBm6p7s454pB8qDPxLlptoB9htNGOdCf0A%2F%2FfgZOumwp7UG%2BZ22%2FAsW5%2Fkk6ZsgN35tsB%2BxBR2piIEtrxNwVwUw91%2Bvx9KobCUr8PRdh8L41KNR6UcRxBtDJE94A%2BNqMOlo2BcSEEJORpIMVAylxdaCT2SiVLiBdjmtJhr644LP7fF1yOI9H2ebwabdZYDe%2F4OLgLgFJPu1dvf%2BZxvbL5Dgfm5CArvdlUXjwO%2FmDP%2BW2x%2Bev49kfkn | url: /plugins/servlet/saml/auth | userName: anonymous
```

### Trying to fix it

I played around with the NameIDFormat a bit, setting it to `transistent` or `persistent`, but nothing did work. I suppose a plain text username is what the plugin actually needs?

I also checked the certificate at least 50 times, as the error message suggests that there's something wrong with the cert/signature. I tried different formats, etc. To no avail.

I'm at a loss here, so I hope that someone might be able to help me or point me in the right direction before I have to dive into the source code.

**Comment by thadig** [comment]
Ok. I got exactly to the same spot. 

Confluence 5.8.14
Confluence SAML 2.0 Plugin 1.0.3 (the one from the marketplace)
IdP: Salesforce

I get the same line in the log:
[bitium.confluence.servlet.SsoLoginServlet] doPost saml plugin error + Validation of protocol message signature failed
 -- referer: (Salesforce URL)  | url: /wiki/plugins/servlet/saml/auth | userName: thomas

From searching the spring.io forum, I start to suspect that the fact that the IdP certificate is a self-signed certificate and thus does not get validated could be the problem. The solution would be to import the self-signed cert into the cacerts store ... something I will try tomorrow.

**Comment by marconett** [comment]
hm. but the SP SHOULD (my understanding of it) just compare the cert that you put into the plugin settings with the incoming response from the idp.

1. sp sends unsigned request
2. idp responds with signed data.
3. sp compares signature with trusted sig (the one in plugin settings).

 it's just used as a signature, not means of encryption.

i couldn't imagine the plugin requiring a cert from a trusted ca without the devs telling us about that requirement. there's also not even any encryption going on, that's why we can read the SAML response in plain xml. so the cert isn't even used for encryption, totally defeting the purpose of a cert from a trusted ca.

**Comment by thadig** [comment]
I agree with you in principle; it should not ... but then, it should work right now, I think, and doesn't.

I tried adding the cert and that did not help. I have been digging through the source a bit and the only place I find where that error could originate is in  http://grepcode.com/file/repo1.maven.org/maven2/org.opensaml/opensaml/2.4.1/org/opensaml/common/binding/security/SAMLProtocolMessageXMLSignatureSecurityPolicyRule.java line 137.

I am not sure yet which criteria fail to evaluate; the only part I found was in:
http://grepcode.com/file/repo1.maven.org/maven2/org.opensaml/opensaml/2.4.1/org/opensaml/common/binding/security/BaseSAMLXMLSignatureSecurityPolicyRule.java#BaseSAMLXMLSignatureSecurityPolicyRule line 67:
new MetadataCriteria(samlContext.getPeerEntityRole(), samlContext.getInboundSAMLProtocol());

I am getting stuck here. Any idea what this policy rule requires?

**Comment by thadig** [comment]
I didn't answer your final question: I had a valid (non-sso) session when that log entry was done.
Without a valid session, it shows anonymous in the log as in your case.

**Comment by marconett** [comment]
Out of desperation, I just now tried the plugin with [LemonLDAP](http://lemonldap-ng.org/), another open source IdP.

By building the plugin manually, I found out, that line 95 in [SsoLoginServlet.java](https://github.com/bitium/confluence-saml-plugin/blob/master/src/main/java/com/bitium/confluence/servlet/SsoLoginServlet.java) (`context.getSamlProcessor().retrieveMessage(messageContext);`) is throwing the exception.

I also came to the same conclusion as @thadig, that line 137 of [SAMLProtocolMessageXMLSignatureSecurityPolicyRule.java](http://grepcode.com/file/repo1.maven.org/maven2/org.opensaml/opensaml/2.4.1/org/opensaml/common/binding/security/SAMLProtocolMessageXMLSignatureSecurityPolicyRule.java) from org.opensaml is where the error occurs.

**Comment by vipinma** [comment]
Any luck on this? I am also getting the same error. 

Confluence 5.9.4
Confluence SAML 2.0 Plugin 1.0.3 (the one from the marketplace)
IdP: WSO2 Identity Server

I am not signing the SAML response from Idp.

If I enable request & response signing validation on Idp then I get error on Idp

org.opensaml.ws.security.SecurityPolicyException: Could not extract the Signature from query string
This is the configuration on Idp side.

![image](https://cloud.githubusercontent.com/assets/17169323/12965403/c3889200-d00a-11e5-815c-9e632c3920a6.png)

Below is configuration on Confluence plug-in
![image](https://cloud.githubusercontent.com/assets/17169323/12965459/3bbd5152-d00b-11e5-80f0-55e7ee7a7595.png)

Any help or pointers will be of great help

**Comment by s-hal** [comment]
You might have already solved it... I had the same problem and the solution for me was to wash the data that was sent in the SAML response. We picked group identity from our AD. In one of the group's name there was the Swedish character ä which caused the error. Renaming the group solved the problem. I assume that the plugin is not 100% UTF-8 compliant.

**Comment by marconett** [comment]
My data source is completely free from any special characters (using a test LDAP with just one user). Same problem.

**Comment by s-hal** [comment]
#facepalm Neglect my post, this was meant for the other SAML plugin.

**Comment by dougmorato** [comment]
@marconett I just tried the bitium plugin today and I have the same exact error: Validation of protocol message signature failed 

I also tried importing the self-signed cert from my IDP  used for signing the SAML assertion but that didn't change a thing, still have the error: Validation of protocol message signature failed .

Found this post and seems to suggest an implementation/change on underlying spring-security: https://stackoverflow.com/questions/23059203/http-status-401-authentication-failed-incoming-saml-message-is-invalid-with

http://forum.spring.io/forum/spring-projects/security/saml/130891-authnresponse-authentication-failed-incoming-saml-message-is-invalid

**Comment by dougmorato** [comment]
+ @thadig

**Comment by yeer** [comment]
@marconett 
I finally solved this problem, you need to modify the `Entity ID`   `SAML 2.0 Plugin configuration`, do not use the default value[ http://domain/confluenceSAML], you need to fill in your IdP Entity ID, http://domain/confluenceSAML this address is for the SP.
![wx20180122-120303](https://user-images.githubusercontent.com/333738/35204902-3c75a080-ff6c-11e7-9a9e-73dee397ff45.png)
